rand_distr/src/zipf.rs

```diff
@@ -0,0 +1,322 @@
+// Copyright 2021 Developers of the Rand project.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! The Zeta and related distributions.
+
+use num_traits::Float;
+use crate::{Distribution, Standard};
+use rand::{Rng, distributions::OpenClosed01};
+use core::fmt;
+
+/// Samples floating-point numbers according to the zeta distribution.
+///
+/// The zeta distribution is a limit of the [`Zipf`] distribution.
+///
+/// # Example
+/// ```
+/// use rand::prelude::*;
+/// use rand_distr::Zeta;
+///
+/// let val: f64 = thread_rng().sample(Zeta::new(1.5).unwrap());
+/// println!("{}", val);
+/// ```
+#[derive(Clone, Copy, Debug)]
+pub struct Zeta<F>
+where F: Float, Standard: Distribution<F>, OpenClosed01: Distribution<F>
+{
+    a_minus_1: F,
+    b: F,
+}
+
+/// Error type returned from `Zeta::new`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ZetaError {
+    /// `a <= 1` or `nan`.
+    ATooSmall,
+}
+
+impl fmt::Display for ZetaError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            ZetaError::ATooSmall => "a <= 1 or is NaN in Zeta distribution",
+        })
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
+impl std::error::Error for ZetaError {}
+
+impl<F> Zeta<F>
+where F: Float, Standard: Distribution<F>, OpenClosed01: Distribution<F>
+{
+    /// Construct a new `Zeta` distribution with given `a` parameter.
+    #[inline]
+    pub fn new(a: F) -> Result<Zeta<F>, ZetaError> {
+        if !(a > F::one()) {
+            return Err(ZetaError::ATooSmall);
+        }
+        let a_minus_1 = a - F::one();
+        let two = F::one() + F::one();
+        Ok(Zeta {
+            a_minus_1,
+            b: two.powf(a_minus_1),
+        })
+    }
+}
+
+impl<F> Distribution<F> for Zeta<F>
+where F: Float, Standard: Distribution<F>, OpenClosed01: Distribution<F>
```

Comment:
Since the distribution is supposed to give integers in 1, 2, ..., should we not implement `Distribution<u64>`, as done in [Geometric](https://docs.rs/rand_distr/0.4.1/rand_distr/struct.Geometric.html)?

Reply:
See below. I think this analogue to `Poisson`, not `Geometric`.

Reply:
Okay! I understand.